### PR TITLE
Cuda backend using shared memory

### PR DESCRIPTION
**Pull request by YohannDudouit**
This backend proposes a Cuda implementation using shared memory.

### CODE REVIEW

File: backends/cuda-shared/ceed-cuda-shared-basis.c [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include <ceed-impl.h>
```
Comment (jeremylt) [inline backends/cuda-shared/ceed-cuda-shared-basis.c]: Backends should *never* use `ceed-impl.h`. Backends should `ceed-backend.h`, and if there is something you need from `ceed-impl.h` we can talk about moving it to `ceed-backend.h`.

File: backends/cuda-shared/ceed-cuda-shared-basis.c [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+#include "../include/ceed.h"
```
Comment (jeremylt) [inline backends/cuda-shared/ceed-cuda-shared-basis.c]: `#include <ceed.h>`

File: backends/cuda-shared/ceed-cuda-shared-basis.c [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+#include "ceed-cuda-shared.h"
+#include "../cuda/ceed-cuda.h"
```
Comment (jeremylt) [inline backends/cuda-shared/ceed-cuda-shared-basis.c]: @jedbrown, I am doing this reference of another backend's header as well in the blocked backend. I am concerned about possible fragility. Thoughts?

File: backends/cuda-shared/ceed-cuda-shared.c [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include <ceed-impl.h>
```
Comment (jeremylt) [inline backends/cuda-shared/ceed-cuda-shared.c]: `ceed-backend.h`

File: backends/cuda-shared/ceed-cuda-shared.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+#include <ceed-backend.h>
+#include "../include/ceed.h"
```
Comment (jeremylt) [inline backends/cuda-shared/ceed-cuda-shared.h]: `ceed.h`

File: backends/cuda-shared/ceed-cuda-shared.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+#include <ceed-impl.h>
```
Comment (jeremylt) [inline backends/cuda-shared/ceed-cuda-shared.h]: remove

**Comment by jeremylt** [comment]
It looks good to me, except for the includes. Would you mind doing a `make style` and `clang-tidy` (#193) pass on the CUDA backends as well?

Does this backend handle your GPU vector test correctly?

File: backends/cuda-shared/ceed-cuda-shared-basis.c [hunk 7]
```diff
@@ -0,0 +1,741 @@
...
+  int ierr;
+  Ceed ceed;
+  ierr = CeedBasisGetCeed(basis, &ceed); CeedChk(ierr);
+  Ceed_Cuda_shared *ceed_Cuda;
+  CeedGetData(ceed, (void *) &ceed_Cuda); CeedChk(ierr);
+  CeedBasis_Cuda_shared *data;
+  CeedBasisGetData(basis, (void *)&data); CeedChk(ierr);
+  const CeedInt transpose = tmode == CEED_TRANSPOSE;
+  // const int optElems[7] = {0,32,8,3,2,1,8};
+  int elemsPerBlock = 1;//basis->Q1d < 7 ? optElems[basis->Q1d] : 1;
```
Comment (jeremylt) [inline backends/cuda-shared/ceed-cuda-shared-basis.c]: future optimization work?

File: backends/cuda-shared/ceed-cuda-shared-basis.c (hunk 3 above)
Comment (jedbrown) [reply backends/cuda-shared/ceed-cuda-shared-basis.c]: I don't think it's a priority, but we may eventually want to discuss how backends can expose a stable interface for extensions.

**Comment by jeremylt** [comment]
Travis is failing due to the MFEM issue. Rebasing will fix it.

File: backends/cuda-shared/ceed-cuda-shared-basis.c (hunk 7 above)
Comment (YohannDudouit) [reply backends/cuda-shared/ceed-cuda-shared-basis.c]: Yes, but I need a proper benchmarking environment to decide which are the best.

**Pull request merged**